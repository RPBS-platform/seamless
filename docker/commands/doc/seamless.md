% seamless(1) Version 0.1 | seamless-cli Documentation

NAME
====

**seamless** - Starts a new Seamless Docker container and runs a command inside it.

SYNOPSIS
========

| **seamless** \[_command_]

DESCRIPTION
===========

Starts a new Seamless Docker container and runs a command inside it.

The current directory is mounted to /cwd, and the command is executed there

/tmp is mounted as well

Uses the host network for the Docker container. Will only work under Linux.

BUGS
====

See GitHub Issues: <https://github.com/sjdv1982/seamless/issues>

AUTHOR
======

Sjoerd de Vries <sjdv1982@gmail.com>

SEE ALSO
========

**seamless-bash(1)**, **seamless-ipython(1)**, **seamless-jupyter(1)**