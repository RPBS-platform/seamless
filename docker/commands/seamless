set -u -e

firstarg=$1 #there must be at least one argument
gdbflags='--cap-add=SYS_PTRACE --security-opt seccomp=unconfined --security-opt apparmor=unconfined'
docker run --rm \
  $gdbflags \
  --network=host \
  -v `pwd`:/cwd \
  -v /tmp:/tmp \
  --workdir /cwd \
  --user $(id -u):$(id -g) \
  rpbs/seamless $*
