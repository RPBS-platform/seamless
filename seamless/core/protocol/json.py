import json
from ...mixed import MixedScalar

def seamless_encoder(obj):
    if isinstance(obj, MixedScalar):
        return default_encoder.encode(obj.value)
    else:
        raise TypeError(repr(obj) + " is not JSON serializable")

def json_encode(obj, *, skipkeys=False, ensure_ascii=True, check_circular=True,
        allow_nan=True, cls=None, indent=None, separators=None,
        default=seamless_encoder, sort_keys=False, **kw):
    """Serialize ``obj`` to a JSON formatted ``str``.

    Wrapper around json.dumps that overrides the default value of "default"
    """
    global default_encoder
    default_encoder = json.JSONEncoder(
      skipkeys=skipkeys,
      ensure_ascii=ensure_ascii,
      check_circular=check_circular,
      allow_nan=allow_nan,
      indent=indent,
      separators=separators,
      default=None,
      sort_keys=sort_keys,
      **kw
    )
    return json.dumps(obj,
      skipkeys=skipkeys,
      ensure_ascii=ensure_ascii,
      check_circular=check_circular,
      allow_nan=allow_nan,
      cls=cls,
      indent=indent,
      separators=separators,
      default=default,
      sort_keys=sort_keys,
      **kw
    )
