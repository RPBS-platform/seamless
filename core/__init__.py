from .cell import cell, pythoncell
from .manager import connect
from .macro import macro
