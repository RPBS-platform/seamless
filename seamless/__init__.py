"""
Seamless: framework for data-driven and live programming
Copyright 2016-2018, Sjoerd de Vries
"""

import sys
import time
import functools
import traceback
import atexit

from abc import abstractmethod
class Wrapper:
    @abstractmethod
    def _unwrap(self):
        pass

from . import silk
from . import mixed

#Dependencies of seamless

# 1. hard dependencies; without these, "import seamless" will fail.
# Still, if necessary, some of these dependencies could be removed, but seamless would have to be more minimalist in loading its lib

import numpy as np

if np.dtype(np.object).itemsize != 8:
    raise ImportError("Seamless requires a 64-bit system")

#silk must be imported before mixed
from . import silk
from . import mixed

from .core import mainloop as _mainloop
from .core.mainloop import mainloop, mainloop_one_iteration, asyncio_finish, workqueue
atexit.register(asyncio_finish)

ipython = None
ipy_error = None
try:
    from IPython import get_ipython
    from IPython.core.error import UsageError
    from IPython.terminal.pt_inputhooks import register as _register_integration_terminal
    from ipykernel.eventloops import register_integration as _register_integration_kernel
except ImportError:
    raise
    ipy_error = "Cannot find IPython"
else:
    ipython = get_ipython()
    if ipython is None:
        ipy_error = "Seamless was not imported inside IPython"

if ipy_error is None:
    last_exception = None
    def new_except_hook(etype, evalue, tb):
        global last_exception
        exc = traceback.format_exception(etype, evalue, tb)
        if exc != last_exception:
            last_exception = exc
            print("".join(exc))

    def patch_excepthook():
        sys.excepthook = new_except_hook
    patch_excepthook()

    def mainloop():
        raise RuntimeError("Cannot run seamless.mainloop() in IPython mode")

    def inputhook_terminal(context):
        while not context.input_is_ready():
            mainloop_one_iteration()
    _register_integration_terminal("seamless", inputhook_terminal)

    @_register_integration_kernel('seamless')
    def inputhook_kernel(kernel):
        while 1:
            t = time.time()
            while time.time() - t < kernel._poll_interval:
                mainloop_one_iteration()        
            kernel.do_one_iteration()

    _register_integration_terminal("seamless", inputhook_terminal)
    
    ipython.enable_gui("seamless")
    
else:
    sys.stderr.write("    " + ipy_error + "\n")
    sys.stderr.write("    Call seamless.mainloop(), seamless.flush() or context.equilibrate() to process cell updates\n")

def flush():
    from .core.mainloop import workqueue
    workqueue.flush()

from .highlevel import *
from .silk import Silk
from .debugger import pdb
