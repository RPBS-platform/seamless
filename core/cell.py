"""Module containing the Cell class."""

import traceback
import inspect
import ast
import os
import copy

from .. import dtypes
from .utils import find_return_in_scope
from .process import Managed
from . import libmanager
from .resource import Resource

class CellLike(object):
    """Base class for cells and contexts
    CellLikes are captured by context.cells"""
    _like_cell = True

class Cell(Managed, CellLike):
    """Default class for cells.

    Cells contain all the state in text form
    """

    class StatusFlags:
        UNINITIALISED, ERROR, OK = range(3)
    StatusFlagNames = ["UNINITIALISED", "ERROR", "OK"]

    _dtype = None
    _data = None  # data, always in text format
    _data_last = None

    _error_message = None
    _status = StatusFlags.UNINITIALISED

    _dependent = False

    _incoming_connections = 0
    _outgoing_connections = 0

    _resource = None

    def __init__(self, dtype, *, naming_pattern="cell"):
        """TODO: docstring."""
        super().__init__()
        from .macro import get_macro_mode
        from .context import get_active_context
        assert dtypes.check_registered(dtype), dtype
        self._dtype = dtype
        self._last_object = None
        self._resource = Resource(self)
        if get_macro_mode():
            ctx = get_active_context()
            ctx._add_new_cell(self, naming_pattern)

    @property
    def resource(self):
        return self._resource

    @property
    def dependent(self):
        """Indicate if the cell is dependent.

        Property is true if the cell has a hard incoming connection,
        e.g. the output of a process.
        """
        return self._dependent

    def set(self, text_or_object):
        """Update cell data from Python code in the main thread."""
        if isinstance(text_or_object, (str, bytes)):
            self._text_set(text_or_object, trusted=False)
        else:
            self._object_set(text_or_object, trusted=False)
        return self

    def fromfile(self, filename):
        return self.resource.fromfile(filename, frames_back=2)

    def _text_set(self, data, trusted):
        try:
            if self._status == self.__class__.StatusFlags.OK \
                    and (data is self._data or data is self._data_last or
                         data == self._data or data == self._data_last):
                return False
        except:
            pass

        try:
            """Check if we can parse the text"""
            dtypes.parse(self._dtype, data, trusted=trusted)

        except dtypes.ParseError:
            self._set_error_state(traceback.format_exc())

            if not trusted:
                raise
        else:
            self._data_last = data
            self._data = data
            self._status = self.__class__.StatusFlags.OK

            if not trusted and self._context is not None:
                manager = self._get_manager()
                manager.update_from_code(self)
        return True

    def _object_set(self, object_, trusted):
        if self._status == self.__class__.StatusFlags.OK \
                and object_ == self._last_object:
            return False
        try:
            """
            Construct the object:
             If the object is already of the correct type,
               then constructed_object is object_
             Some datatypes (i.e. silk) can construct the object from
              heterogenous input
            """
            dtypes.construct(self._dtype, object_)

        except dtypes.ConstructionError:
            self._set_error_state(traceback.format_exc())

            if not trusted:
                raise
        else:
            data = dtypes.serialize(self._dtype, object_)
            # Normally no error here...
            self._data = data
            self._status = self.__class__.StatusFlags.OK
            self._last_object = copy.deepcopy(object_)

            if not trusted and self._context is not None:
                manager = self._get_manager()
                manager.update_from_code(self)
        return True

    def touch(self):
        if self._status != self.__class__.StatusFlags.OK:
            return
        if self._context is not None:
            manager = self._get_manager()
            manager.update_from_code(self)

    def _update(self, data):
        """Invoked when cell data is updated by a process."""
        return self._text_set(data, trusted=True)

    def connect(self, target):
        """Connect the cell to a process's input pin."""
        manager = self._get_manager()
        manager.connect(self, target)

    @property
    def dtype(self):
        """The cell's data type."""
        return self._dtype

    @property
    def data(self):
        """The cell's data in text format."""
        return copy.deepcopy(self._data)

    @property
    def status(self):
        """The cell's current status."""
        return self.StatusFlagNames[self._status]

    @property
    def error_message(self):
        """The cell's current error message.

        Returns None is there is no error
        """
        return self._error_message

    def _on_connect(self, pin, process, incoming):
        from .process import OutputPinBase
        if incoming:
            if self._dependent and isinstance(pin, OutputPinBase):
                raise Exception(
                 "Cell is already the output of another process"
                )
            if isinstance(pin, OutputPinBase):
                self._dependent = True
            self._incoming_connections += 1
        else:
            self._outgoing_connections += 1

    def _on_disconnect(self, pin, process, incoming):
        from .process import OutputPinBase
        if incoming:
            if isinstance(pin, OutputPinBase):
                self._dependent = False
            self._incoming_connections -= 1
        else:
            self._outgoing_connections -= 1

    def _set_error_state(self, error_message=None):
        if error_message is not None:
            self._status = self.StatusFlags.ERROR
            if error_message != self._error_message:
                print(error_message)
        self._error_message = error_message

    def add_macro_object(self, macro_object, macro_arg):
        manager = self._get_manager()
        manager.add_macro_listener(self, macro_object, macro_arg)

    def remove_macro_object(self, macro_object, macro_arg):
        manager = self._get_manager()
        manager.remove_macro_listener(self, macro_object, macro_arg)

    def destroy(self):
        if self._destroyed:
            return
        #print("CELL DESTROY", self)
        self.resource.destroy()
        super().destroy()

class PythonCell(Cell):
    """
    A cell containing Python code.

    Python cells may contain either a code block or a function
    Processes that are connected to it may require either a code block or a
     function
    Mismatch between the two is not a problem, unless:
          Connected processes have conflicting block/function requirements
        OR
            A function is required (typically, true for transformers)
          AND
            The cell contains a code block
          AND
            The code block contains no return statement
    """

    class CodeTypes:
        ANY, FUNCTION, BLOCK = range(3)

    _dtype = ("text", "python")

    _code_type = CodeTypes.ANY
    _required_code_type = CodeTypes.ANY

    def _text_set(self, data, trusted):
        if data == self._data:
            return False
        try:
            """Check if the code is valid Python syntax"""
            ast_tree = compile(data, str(self), "exec", ast.PyCF_ONLY_AST)

        except SyntaxError:
            if not trusted:
                raise

            else:
                self._set_error_state(traceback.format_exc())

        else:
            is_function = (
             len(ast_tree.body) == 1 and
             isinstance(ast_tree.body[0], ast.FunctionDef)
            )

            # If this cell requires a function, but wasn't provided
            #  with a def block
            if not is_function and \
                    self._required_code_type == self.CodeTypes.FUNCTION:
                # Look for return node in AST
                try:
                    find_return_in_scope(ast_tree)
                except ValueError:
                    exception = SyntaxError(
                     "Block must contain return statement(s)"
                    )

                    if trusted:
                        self._set_error_state("{}: {}".format(
                         exception.__class__.__name__, exception.msg)
                        )
                        return

                    else:
                        raise exception

            self._data = data
            self._code_type = self.CodeTypes.FUNCTION if is_function else \
                self.CodeTypes.BLOCK
            self._set_error_state(None)
            self._status = self.StatusFlags.OK

            if not trusted and self._context is not None:
                manager = self._get_manager()
                manager.update_from_code(self)
            return True

    def _object_set(self, object_, trusted):
        from .utils import strip_source
        try:
            """
            Try to retrieve the source code
            Will only work if code is a function
            """
            if not inspect.isfunction(object_):
                raise Exception("Python object must be a function")

            code = inspect.getsource(object_)
            code = strip_source(code)

        except:
            self._set_error_state(traceback.format_exc())

            if not trusted:
                raise

        else:
            self._code_type = self.CodeTypes.FUNCTION
            oldcode = self._data
            self._data = code
            self._status = self.__class__.StatusFlags.OK

            if not trusted and self._context is not None:
                manager = self._get_manager()
                manager.update_from_code(self)
            return code != oldcode

    def _on_connect(self, pin, process, incoming):
        exc1 = """Cannot connect to %s: process requires a code function
        whereas other connected processes require a code block"""
        exc2 = """Cannot connect to %s: process requires a code block
        whereas other connected processes require a code function"""

        if not incoming:
            if self._required_code_type == self.CodeTypes.BLOCK and \
                    process._required_code_type == self.CodeTypes.FUNCTION:
                raise Exception(exc1 % type(process))
            elif self._required_code_type == self.CodeTypes.FUNCTION and \
                    process._required_code_type == self.CodeTypes.BLOCK:
                raise Exception(exc2 % type(process))

        Cell._on_connect(self, pin, process, incoming)
        if not incoming:
            self._required_code_type = process._required_code_type

    def _on_disconnect(self, pin, process, incoming):
        Cell._on_disconnect(self, pin, process, incoming)
        if self._outgoing_connections == 0:
            self._required_code_type = self.CodeTypes.ANY

_handlers = {
    ("text", "code", "python"): PythonCell
}


def cell(dtype):
    """Factory function for a Cell object."""
    cell_cls = Cell
    if dtype in _handlers:
        cell_cls = _handlers[dtype]

    newcell = cell_cls(dtype)
    return newcell


def pythoncell():
    """Factory function for a PythonCell object."""
    return cell(("text", "code", "python"))
