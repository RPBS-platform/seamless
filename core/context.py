"""Module for Context class."""
from weakref import WeakValueDictionary
from . import SeamlessBase
from .cell import Cell, CellLike, ExportedCell
from .process import Managed, Process, ProcessLike, InputPinBase, \
  ExportedInputPin, OutputPinBase, ExportedOutputPin, EditorOutputPin
from contextlib import contextmanager as _pystdlib_contextmanager
_active_context = None

#TODO: subcontexts inherit manager from parent? see process.connect source code

def set_active_context(ctx):
    global _active_context
    assert ctx is None or isinstance(ctx, Context)
    _active_context = ctx

def get_active_context():
    return _active_context

@_pystdlib_contextmanager
def active_context_as(ctx):
    previous_context = get_active_context()
    try:
        set_active_context(ctx)
        yield
    finally:
        set_active_context(previous_context)

class Context(SeamlessBase, CellLike, ProcessLike):
    """Context class. Organizes your cells and processes hierarchically.
    """

    _name = None
    _registrar = None
    _like_cell = False          #can be set to True by export
    _like_process = False       #can be set to True by export
    _children = None
    _childids = None
    _manager = None
    registrar = None
    _pins = None
    _auto = None

    def __init__(
        self,
        name=None,
        context=None,
        active_context=True,
    ):
        """Construct a new context.

        Args:
            context (optional): parent context
            active_context (default: True): Sets the newly constructed context
                as the active context. Subcontexts constructed by macros are
                automatically parented to the active context
        """
        n = name
        if context is not None and context._name is not None:
            n = context._name + "." + str(n)
        self._name = name
        self._pins = {}
        self._children = {}
        self._childids = WeakValueDictionary()
        self._auto = set() #TODO: save this also when serializing
        if context is not None:
            self._manager = context._manager
        else:
            from .process import Manager
            self._manager = Manager()
        if active_context:
            set_active_context(self)
        from .registrar import RegistrarAccessor
        self.registrar = RegistrarAccessor(self)

    _dir = ["_name", "export", "registrar", "cells"]

    @property
    def cells(self):
        from .cell import CellLike
        return [v for k,v in self._children.items() if isinstance(v, CellLike) and v._like_cell]

    def __dir__(self):
        return [c for c in self._children.keys() if c not in self._auto] \
         + [c for c in self._pins.keys()] + self._dir

    def _add_child(self, childname, child):
        self._children[childname] = child
        self._childids[id(child)] = child
        child._set_context(self)

    def _add_new_cell(self, cell, naming_pattern="cell"):
        from .cell import Cell
        assert isinstance(cell, Cell)
        count = 0
        while 1:
            count += 1
            cell_name = naming_pattern + str(count)
            if not self._hasattr(cell_name):
                break
        self._auto.add(cell_name)
        self._add_child(cell_name, cell)
        return cell

    def _add_new_process(self, process, naming_pattern="process"):
        from .process import Process
        assert isinstance(process, Process)
        count = 0
        while 1:
            count += 1
            process_name = naming_pattern + str(count)
            if not self._hasattr(process_name):
                break
        self._auto.add(process_name)
        self._add_child(process_name, process)
        return process

    def _new_subcontext(self, naming_pattern="ctx"):
        count = 0
        while 1:
            count += 1
            context_name = naming_pattern + str(count)
            if not self._hasattr(context_name):
                break
        ctx = context(context=self, active_context=False)
        self._auto.add(context_name)
        self._add_child(context_name, ctx)
        return ctx

    def __setattr__(self, attr, value):
        if hasattr(self.__class__, attr):
            return object.__setattr__(self, attr, value)
        if attr in self._pins:
            raise AttributeError(
             "Cannot assign to pin ''%s'" % attr)
        if attr in self._children and self._children[attr] is not value:
            self._children[attr].destroy()

        assert isinstance(value, (Managed, CellLike, ProcessLike)), type(value)
        value._set_context(self, force_detach=True)
        self._add_child(attr, value)

    def __getattr__(self, attr):
        if attr in self._pins:
            return self._pins[attr]
        elif attr in self._children:
            return self._children[attr]
        else:
            raise AttributeError(attr)

    def __delattr__(self, attr):
        if attr in self._pins:
            raise AttributeError("Cannot delete pin: '%s'" % attr)
        elif attr not in self._children:
            raise AttributeError(attr)
        child = self._children[attr]
        child.destroy()
        self._children.pop(attr)

    def _hasattr(self, attr):
        if hasattr(self.__class__, attr):
            return True
        if attr in self._children:
            return True
        if attr in self._pins:
            return True
        return False

    def export(self, child, forced=[]):
        """Exports all unconnected inputs and outputs of a child

        If the child is a cell (or cell-like context):
            - export the child's input as primary input (if unconnected)
            - export the child's output as primary output (if unconnected)
            - export any other pins, if forced
            - sets the context as cell-like
        If the child is a process (or process-like context):
            - export all unconnected input and output pins of the child
            - export any other pins, if forced
            - sets the context as process-like

        Arguments:

        child: a direct or indirect child (grandchild) of the context
        forced: contains a list of pin names that are exported in any case
          (even if not unconnected).
          Use "_input" and "_output" to indicate primary cell input and output

        """
        assert child.context._part_of(self)
        mode = None
        if isinstance(child, CellLike) and child._like_cell:
            mode = "cell"
            pins = ["_input", "_output"]
        elif isinstance(child, ProcessLike) and child._like_process:
            mode = "process"
            pins = child._pins.keys()
        else:
            raise TypeError(child)

        def is_connected(pinname):
            if isinstance(child, CellLike) and child._like_cell:
                child2 = child
                if not isinstance(child, Cell):
                    child2 = child.get_cell()
                if pinname == "_input":
                    return (child2._incoming_connections > 0)
                elif pinname == "_output":
                    return (child2._outgoing_connections > 0)
                else:
                    raise ValueError(pinname)
            else:
                pin = child._pins[pinname]
                if isinstance(pin, InputPinBase):
                    manager = pin._get_manager()
                    con_cells = manager.pin_to_cells.get(pin.get_pin_id(), [])
                    return (len(con_cells) > 0)
                elif isinstance(pin, OutputPinBase):
                    pin = pin.get_pin()
                    if isinstance(pin, EditorOutputPin):
                        return (len(pin.solid._cell_ids) > 0) or \
                         (len(pin.liquid._cell_ids) > 0)
                    else:
                        return (len(pin._cell_ids) > 0)
                else:
                    raise TypeError(pin)
        pins = [p for p in pins if not is_connected(p)] + forced
        if not len(pins):
            raise Exception("Zero pins to be exported!")
        for pinname in pins:
            if self._hasattr(pinname):
                raise Exception("Cannot export pin '%s', context has already this attribute" % pinname)
            if isinstance(child, CellLike) and child._like_cell:
                if not isinstance(child, Cell):
                    child = child.get_cell()
                    self._pins[pinname] = ExportedCell(child)
            else:
                pin = child._pins[pinname]
                if isinstance(pin, InputPinBase):
                    self._pins[pinname] = ExportedInputPin(pin)
                elif isinstance(pin, OutputPinBase):
                    self._pins[pinname] = ExportedOutputPin(pin)

        if mode == "cell":
            self._like_cell = True
        elif mode == "process":
            self._like_process = True

    def _part_of(self, ctx):
        assert isinstance(ctx, Context)
        if ctx is self:
            return True
        elif self._context is None:
            return False
        else:
            return self._context._part_of(ctx)

    def _root(self):
        if self._context is None:
            return self
        else:
            return self._context._root()

    def _set_context(self, ctx, force_detach=True):
        if self._context is not None:
            if self._context is not ctx or force_detach:
                for childname, child in self._context._children.items():
                    if child is self:
                        self._context._children.pop(childname)
                        break
                else:
                    print("WARNING, orphaned context child?")
        self._context = ctx

    def destroy(self):
        if self._destroyed:
            return
        #print("CONTEXT DESTROY")
        for childname in list(self._children):
            child = self._children[childname]
            child.destroy()
        super().destroy()

def context(**kwargs):
    """Return a new Context object."""
    ctx = Context(**kwargs)
    return ctx
