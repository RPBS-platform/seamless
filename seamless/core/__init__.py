import weakref
from enum import Enum

class IpyString(str):
    def _repr_pretty_(self, p, cycle):
        return p.text(str(self))

class SeamlessBase:
    _destroyed = False
    _macro_object = None # macro object that CREATED this instance
    _context = None
    _last_context = None
    name = None

    StatusFlags = Enum('StatusFlags', ('OK', 'PENDING', 'UNDEFINED', 'UNCONNECTED', 'ERROR'))
    _status = StatusFlags.UNDEFINED

    @property
    def path(self):
        if self._context is None:
            return ()
        elif self._context.path is None:
            return ("<None>", self.name)
        else:
            return self._context.path + (self.name,)

    def _validate_path(self, required_path=None):
        if required_path is None:
            required_path = self.path
        else:
            assert self.path == required_path, (self.path, required_path)
        return required_path


    def _set_context(self, context, name, force_detach=False):
        from .context import Context
        assert isinstance(context, Context)
        if self._context is not None:
            assert self.name is not None
            if context is not self._context or force_detach:
                childname = self.name
                assert self._context._children[childname] is self
                self._context._children.pop(childname)
                if childname in self._context._auto:
                    self._context._auto.remove(childname)
        if context is not None:
            self._last_context = context
        self._context = context
        self.name = name
        return self

    @property
    def context(self):
        return self._context


    def _macro_control(self):
        if self._macro_object is not None:
            return self

        ret = None
        if self.context is not None:
            ret = self.context._macro_control()
        else:
            return None

    def format_path(self):
        if self.path is None:
            ret = "<None>"
        else:
            ret = "." + ".".join(self.path)
        return ret

    def __str__(self):
        ret = "Seamless object: " + self.format_path()
        return ret

    def __repr__(self):
        return self.__str__()

    @property
    def macro(self):
        """Returns the macro object associated with this construct"""
        return self._macro_object
    def _set_macro_object(self, macro_object):
        self._macro_object = macro_object

#from .cell import Cell ###
#from .worker import Worker ###
from .context import Context, context ###
