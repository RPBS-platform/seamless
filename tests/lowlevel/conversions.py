import asyncio
import json
from seamless.core import context, cell

ctx = context(toplevel=True)

value = [42, "test", {"mykey": "myvalue"}, True]
value_json = json.dumps(value)

import time
start_time = time.time()
print_ORIGINAL = print

def print(*args):
    if not len(args):
        print_ORIGINAL()
        return
    args = [str(arg) for arg in args]
    elapsed_time = time.time() - start_time
    print_ORIGINAL("Time: %.1f ms," % (1000 * elapsed_time), *args)

ctx.txt1 = cell("text")
ctx.txt2 = cell("text")
ctx.txt3 = cell("text")
ctx.txt4 = cell("text")
ctx.plain = cell("plain")
ctx.plain2 = cell("plain")
ctx.mixed = cell("mixed")

print("*** Start ***")
print(ctx.txt1.value)
print()

print("*** Stage 0 ***")
ctx.txt1.set(value_json)
print()

print("*** Stage 1a ***")
ctx.compute()
value = ctx.txt1.data
print("ctx.txt1", value)

print("*** Stage 1b ***")
ctx.txt2.set(value)
ctx.txt3.set(value)

print("*** Stage 1c ***")
ctx.compute()
print("ctx.txt2", ctx.txt2.data)
print("ctx.txt3", ctx.txt3.data)
print()

print("*** Stage 2 ***")
ctx.txt3.connect(ctx.txt4)
ctx.compute()
print("ctx.txt4", ctx.txt4.data)
print()

print("*** Stage 3 ***")
ctx.txt1.connect(ctx.plain)
ctx.compute()
print("ctx.plain", ctx.plain.data)

print("*** Stage 3a ***")
ctx.txt3.connect(ctx.plain2)
ctx.compute()
print("ctx.plain2", ctx.plain2.data)

print("*** Stage 3b ***")
ctx.txt3.connect(ctx.mixed)
ctx.compute()
print("ctx.mixed", ctx.mixed.data)

print("*** Stage 4 ***")
ctx.ipy = cell("ipython").set("""
%%timeit
x = 42

""")
ctx.py = cell("python")
ctx.ipy.connect(ctx.py)
ctx.compute()
print("ctx.py", ctx.py.data)
print()

print("*** Stage 5 ***")
ctx.plain3 = cell("plain").set("Test string!!")
ctx.text = cell("text")
ctx.plain3.connect(ctx.text)
ctx.str = cell("str")
ctx.plain3.connect(ctx.str)
ctx.compute()
print("ctx.plain3", ctx.plain3.data, ctx.plain3.buffer)
print("ctx.text", ctx.text.data, ctx.text.buffer)
print("ctx.str", ctx.str.data, ctx.str.buffer)
