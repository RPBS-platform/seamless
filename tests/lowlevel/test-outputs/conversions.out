Time: 0.3 ms, *** Start ***
Time: 1.0 ms, None

Time: 1.0 ms, *** Stage 0 ***

Time: 1.4 ms, *** Stage 1a ***
Time: 3.3 ms, ctx.txt1 [42, "test", {"mykey": "myvalue"}, true]
Time: 3.3 ms, *** Stage 1b ***
Time: 3.5 ms, *** Stage 1c ***
Time: 5.3 ms, ctx.txt2 [42, "test", {"mykey": "myvalue"}, true]
Time: 5.8 ms, ctx.txt3 [42, "test", {"mykey": "myvalue"}, true]

Time: 5.8 ms, *** Stage 2 ***
Time: 208.5 ms, ctx.txt4 [42, "test", {"mykey": "myvalue"}, true]

Time: 208.7 ms, *** Stage 3 ***
Time: 420.0 ms, ctx.plain [42, "test", {"mykey": "myvalue"}, true]
Time: 420.2 ms, *** Stage 3a ***
Time: 620.6 ms, ctx.plain2 [42, "test", {"mykey": "myvalue"}, true]
Time: 620.8 ms, *** Stage 3b ***
Time: 821.2 ms, ctx.mixed [42, "test", {"mykey": "myvalue"}, true]
Time: 821.3 ms, *** Stage 4 ***
Time: 3501.8 ms, ctx.py get_ipython().run_cell_magic('timeit', '', 'x = 42')

Time: 3501.9 ms, *** Stage 5 ***
Time: 3700.3 ms, ctx.plain3 Test string!! b'"Test string!!"\n'
Time: 3701.0 ms, ctx.text Test string!! b'Test string!!\n'
Time: 3701.6 ms, ctx.str Test string!! b'"Test string!!"\n'
